src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/14.diff.sql

```diff
@@ -0,0 +1,542 @@
+      CONSTRAINT UQ_ResourceChangeType_Name UNIQUE NONCLUSTERED 
+     (
+         Name ASC
+     )WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
+     ) ON [PRIMARY]
+
+     SET IDENTITY_INSERT [dbo].[ResourceChangeType] ON 
+
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (0, N'Creation')
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (1, N'Update')
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (2, N'Deletion')
+
+     SET IDENTITY_INSERT dbo.ResourceChangeType OFF
+END
+GO
+
+/*************************************************************
+    Stored procedures for capturing and fetching resource changes
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     CaptureResourceChanges
+--
+-- DESCRIPTION
+--     Inserts resource change data
+--
+-- PARAMETERS
+--     @isDeleted
+--         * Whether this resource marks the resource as deleted.
+--     @version
+--         * The version of the resource being written
+--     @resourceId
+--         * The resource ID
+--     @resourceTypeId
+--         * The ID of the resource type
+--
+-- RETURN VALUE
+--     It does not return a value.
+--
+CREATE OR ALTER PROCEDURE dbo.CaptureResourceChanges
+    @isDeleted bit, 
+    @version int, 
+    @resourceId varchar(64),
+    @resourceTypeId smallint
+AS
+BEGIN
+    DECLARE @changeType SMALLINT
+    IF (@isDeleted = 1) BEGIN
+        SET @changeType = 2    /* DELETION */
+    END
+    ELSE BEGIN
+        IF (@version = 1) BEGIN
+            SET @changeType = 0 /* CREATION */
+        END
+        ELSE BEGIN
+            SET @changeType = 1 /* UPDATE */
+        END
+    END
+
+    INSERT INTO dbo.ResourceChangeData
+        (ResourceId, ResourceTypeId, ResourceVersion, ResourceChangeTypeId)
+    VALUES
+        (@resourceId, @resourceTypeId, @version, @changeType)
+END
+GO
+
+--
+-- STORED PROCEDURE
+--     FetchResourceChanges
+--
+-- DESCRIPTION
+--     Returns the number of resource change records from startId. The start id is inclusive.
+--
+-- PARAMETERS
+--     @startId
+--         * The start id of resource change records to fetch.
+--     @pageSize
+--         * The page size for fetching resource change records.
+--
+-- RETURN VALUE
+--     Resource change data rows.
+--
+CREATE OR ALTER PROCEDURE dbo.FetchResourceChanges
+    @startId bigint, 
+    @pageSize int
+AS
+BEGIN
+
+    SET NOCOUNT ON;
+
+    SET TRANSACTION ISOLATION LEVEL REPEATABLE READ;
+
+    SELECT TOP(@pageSize) c.Id
+      ,c.Timestamp
+      ,c.ResourceId
+      ,c.ResourceTypeId
+      ,t.Name AS ResourceTypeName
+      ,c.ResourceVersion
+      ,c.ResourceChangeTypeId
+      FROM dbo.ResourceChangeData c
+      LEFT JOIN dbo.ResourceType t ON c.ResourceTypeId = t.ResourceTypeId
```

Comment:
What will we do if ResourceTypeId is missing from dbo.ResourceType?  Is that an expected scenario?

Reply:
The dbo.ResourceType is a reference table. Unlikely ResourceTypeId will be missing. But, if it is, the resource type name will be empty on the event payload.

Reply:
Are we going to document that behavior and what customers should do in that case?  The left join implies that it's possible for the resource type name to be missing.  If we don't intend for that to happen, then perhaps something should prevent/handle it.

Reply:
Let me take back my last statement. Given the fact that the resource type is one of the important factors in the Fhir application and the resource change data is inserted during upsertResource transactions, resource type id will be a valid value. And, to ensure to have a valid resource type name on even payload, adding logic in the publish event service to check if resource name and resource Id are not empty. If they are empty, log exceptions to investigate a problem, and skip that record to send an event since the record is invalid.

Reply:
Makes sense, thanks.  Since the ResourceType table rarely changes, another option is to read the table once into memory in the eventing processor, map the id to the name there, and skip the join here in the sproc.  Thoughts?  We would need to make sure the table is reloaded if the FHIR service adds new resource types.

Reply:
We had a discussion about caching resource types at the startup of an event agent service. Based on the fact that resource types are not changed during runtime, removing the left join of the resource change data with a resource type table from the select statement to simplify the query execution plan.